docs/developers/architecture/magicgui_type_reg.md

```diff
@@ -0,0 +1,69 @@
+[type hints](https://peps.python.org/pep-0484/) to infer the appropriate widget type
+for a given function parameter. It allows third party packages
+(like `napari`) to [register](https://pyapp-kit.github.io/magicgui/type_map/#registering-support-for-custom-types) support for their types using
+{func}`~magicgui.type_map.register_type`. `napari` registers
+a number of types, additionally specifying, where appropriate:
+
+* widget type for the parameter type
+* function for updating inputs for the widget
+* return callback, for return types
+
+This enables `magicgui` widgets to be easily created via type annotations.
+
+```{note}
+This page provides implementation details on `napari`-specific type registration
+in `magicgui` and is aimed at `napari` developers.
+
+For information about using `magicgui` (for users and plugin developers) see
+[](creating-widgets).
+```
+
+## Registration details
+
+`napari` types are either registered via the
+{func}`@register_type <magicgui.type_map.register_type>` decorator when they are defined or in
+[`napari/types.py`](https://github.com/napari/napari/blob/main/napari/types.py).
+For the full list of types registered, see [](magicgui-parameter-annotations).
+
+All 'layer' types provide a `choices` callable when registering.
+This means that annotating with these types creates an
+{class}`~magicgui.widgets.bases.CategoricalWidget`, which will get updated via the
+`choices` callable. This callable is either `get_layers_data` or `get_layers`.
+These functions retrieve the closest parent `Viewer` of the native
+{class}`~magicgui.widgets.bases.CategoricalWidget` widget and returns a list of
+{class}`~napari.layers.Layer` or tuple of format ('layer name', `<LayerType>Data`).
+This callable is set to the `choices` attribute of the
+{class}`~magicgui.widgets.bases.CategoricalWidget` in its
+{meth}`~magicgui.widgets.bases.CategoricalWidget.reset_choices` method and thus
+gets called via {class}`~magicgui.widgets.bases.CategoricalWidget`'s `choices` setter.
+`napari` {meth}`~napari.qt.Window.add_dock_widget` checks if the dock widget has
+a `reset_choices` attribute and if so, connects it to layer events.
+Note that `magicgui` {class}`~magicgui.widgets.bases.ContainerWidget`'s will call
+`reset_choices` on all subwidgets. This means that when the dock widget is a
+{class}`~magicgui.widgets.bases.ContainerWidget`, any subwidgets (e.g.,
+{class}`~magicgui.widgets.bases.CategoricalWidget`s) are updated
+whenever layers change.
+**Note that {meth}`~napari.qt.Window.add_dock_widget` will
+connect any existing `reset_choices` widget attribute to layer events for all widgets,
+not just `magicgui` widgets.**
+
+The 'layer' types also specify a `return_callback` function that adds the layer
+to the closest parent `Viewer` of the native widget when a 'layer' type is a return
+annotation.
+
+{class}`~napari.viewer.Viewer` differs from the Layer types. `napari` simply specifies
```

Comment:
Is the `Viewer` registered with magicgui at all? If not then I think we should explicitly state that here

Reply:
It is registered and it is super weird ([ref](https://github.com/napari/napari/blob/cf68ee5ffbe6f5ff4861642c759eaff181ef6749/napari/viewer.py#L15) ):
```python
@mgui.register_type(bind=_magicgui.proxy_viewer_ancestor)
class Viewer(ViewerModel):
```

what 'bind' does ([ref](https://github.com/pyapp-kit/magicgui/blob/1e78e6f0bee37aeb9ce7d300a52f5f120141f834/src/magicgui/type_map/_type_map.py#L423C1-L426C68)):

```python
    elif "bind" in _options:
        # if we're binding a value to this parameter, it doesn't matter what type
        # of ValueWidget is used... it usually won't be shown
        _TYPE_DEFS[resolved_type] = (widgets.EmptyWidget, _options)
```

`EmptyWidget` is just a 'hidden' widget, the user has to specify widget they want for `Viewer` annotation. I don't know how a `Viewer` would be used as an input widget though..?!